src/cockpit/polkit.py

```diff
@@ -0,0 +1,155 @@
+
+logger = logging.getLogger(__name__)
+
+Identity = Tuple[str, Dict[str, Variant]]
+
+
+# https://www.freedesktop.org/software/polkit/docs/latest/eggdbus-interface-org.freedesktop.PolicyKit1.AuthenticationAgent.html
+
+# Note that we don't implement the CancelAuthentication() API. pkexec gets called in a way that has no opportunity to
+# cancel an ongoing authentication from the pkexec side. On the UI side cancellation is implemented via the standard
+# asyncio process mechanism. If we ever need CancelAuthentication(), we could keep a cookie → get_current_task()
+# mapping, but that method is not available for Python 3.6 yet.
+
+class org_freedesktop_PolicyKit1_AuthenticationAgent(bus.Object):
+    def __init__(self, responder: InteractionResponder):
+        super().__init__()
+        self.responder = responder
+
+    # confusingly named: this actually does the whole authentication dialog, see docs
+    @bus.Interface.Method('', ['s', 's', 's', 'a{ss}', 's', 'a(sa{sv})'])
+    async def begin_authentication(self, action_id: str, message: str, icon_name: str,
+                                   details: Dict[str, str], cookie: str, identities: List[Identity]):
+        logger.debug('BeginAuthentication: action %s, message "%s", icon %s, details %s, cookie %s, identities %r',
+                     action_id, message, icon_name, details, cookie, identities)
+        # only support authentication as ourselves, as we don't yet have the
+        # protocol plumbing and UI to select an admin user
+        my_uid = os.geteuid()
+        for (auth_type, subject) in identities:
+            if auth_type == 'unix-user' and 'uid' in subject and subject['uid'].value == my_uid:
+                logger.debug('Authentication subject %s matches our uid %d', subject, my_uid)
+                break
+        else:
+            logger.warning('Not supporting authentication as any of %s', identities)
+            return
+
+        user_name = pwd.getpwuid(my_uid).pw_name
+        process = await asyncio.create_subprocess_exec(HELPER_PATH, user_name, cookie,
+                                                       stdin=asyncio.subprocess.PIPE,
+                                                       stdout=asyncio.subprocess.PIPE)
+        try:
+            await self._communicate(process)
+        except asyncio.CancelledError:
+            logger.debug('Cancelled authentication')
+            process.terminate()
+        finally:
+            res = await process.wait()
+            logger.debug('helper exited with code %i', res)
+
+    async def _communicate(self, process: asyncio.subprocess.Process):
+        assert process.stdin
+        assert process.stdout
+
+        while process.returncode is None:
+            line = await process.stdout.readline()
+            logger.debug('Read line from helper: %s', line)
+
+            # usually: PAM_PROMPT_ECHO_OFF Password: \n
+            if line.startswith(b'PAM_PROMPT'):
```

Comment:
Instead of all of the `.startswith()` trying to

```python
    command, _, args = line.partition(b' ')
```

and using a usual `if`/`elif` on the value of `command`?  You might also do the `.decode()` at the toplevel (before the partition) to avoid having to deal with bytestrings below.

Does the async subprocess code not do `text=True`?